osmosis-labs/osmosis
# Dockerfile improvements

Closes: 
- https://github.com/osmosis-labs/osmosis/issues/1680
- https://github.com/osmosis-labs/osmosis/issues/1682
- https://github.com/osmosis-labs/osmosis/issues/1681

## What is the purpose of the change

This  PR improves our `Dockerfile` by:

- Populating the `.dockerignore` with more folders. 
- Reducing the number of layers 
- Fixing a bug: `$WASMVM` was hardcoded for the checksum

Now builds are significantly faster due to better use of layer caching and previous work by @ValarDragon on the `go build` `go mod download` cache.

I have also added the relevant commands in the `Makefile`:

- `make docker-build` builds local docker image using the default `distroless` runner image 
- `make docker-build-distroless` same as above
- `make docker-build-alpine` builds local docker image using the `alpine` runner image 
- `make docker-build-nonroot` builds local docker image using the `distroless-nonroot` runner image 

## Brief Changelog

- Reduce the number of layers in `Dockerfile`
- Extend `.dockerignore`

## Testing and Verifying

1. Build image: 

```bash
docker build -t osmosis:local .
```

2. Run image:

```bash
docker run -ti osmosis:local version
```

You should see an empty line as output because there is no version set.
To set a version:

```bash
docker build \
  --build-arg GIT_VERSION="my-version" \
  --build-arg GIT_COMMIT="my-commit" \
  -t osmosis:local 
  . 
```

To get the correct values programmatically:

```
VERSION=$(echo $(shell git describe --tags) | sed 's/^v//')
COMMIT= $(git log -1 --format='%H')

docker build \
  --build-arg GIT_VERSION=$VERSION \
  --build-arg GIT_COMMIT=$COMMIT \
  -t osmosis:local 
  . 
```

> The `--build-arg` shouldn't be needed for building an image for testing purposes.
>
> Here I'm trying to be more verbose in explaining the internals

or more simply:

1. Build:

```
make docker-build
```

2. Run:

```bash
docker run -ti osmosis:local version
```

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable
